src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +238,331 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
-                    newIndex++;
-                    goto nextNewChange;
-                }
             }
-            else
+
+            // there may be remaining old changes or remaining new changes (not both, and not neither)
+            switch (oldIndex == oldChanges.Length, newIndex == newChanges.Length)
+            {
+                case (true, true):
+                case (false, false):
+                    throw new InvalidOperationException();
+            }
+
+            while (oldIndex < oldChanges.Length)
+            {
+                addAndAdjustOldDelta(builder, ref oldDelta, oldChange);
+                tryGetNextOldChange();
+            }
+
+            while (newIndex < newChanges.Length)
             {
-                // no more new changes, just add remaining old changes
-                while (oldIndex < oldChanges.Length)
+                addAdjustedNewChange(builder, oldDelta, newChange);
+                tryGetNextNewChange();
+            }
+
+            return builder.ToImmutableAndFree();
+
+            bool tryGetNextOldChange()
+            {
+                oldIndex++;
+                if (oldIndex < oldChanges.Length)
+                {
+                    oldChange = oldChanges[oldIndex];
+                    return true;
+                }
+                else
                 {
-                    AddRange(list, oldChanges[oldIndex]);
-                    oldIndex++;
+                    oldChange = default;
+                    return false;
                 }
             }
 
-            return list.ToImmutableArray();
-        }
-
-        private static void AddRange(List<TextChangeRange> list, TextChangeRange range)
-        {
-            if (list.Count > 0)
+            bool tryGetNextNewChange()
             {
-                var last = list[list.Count - 1];
-                if (last.Span.End == range.Span.Start)
+                newIndex++;
+                if (newIndex < newChanges.Length)
                 {
-                    // merge changes together if they are adjacent
-                    list[list.Count - 1] = new TextChangeRange(new TextSpan(last.Span.Start, last.Span.Length + range.Span.Length), last.NewLength + range.NewLength);
-                    return;
+                    newChange = new UnadjustedNewChange(newChanges[newIndex]);
+                    return true;
                 }
                 else
                 {
-                    Debug.Assert(range.Span.Start > last.Span.End);
+                    newChange = default;
+                    return false;
                 }
             }
 
-            list.Add(range);
+            static void addAndAdjustOldDelta(ArrayBuilder<TextChangeRange> builder, ref int oldDelta, TextChangeRange oldChange)
+            {
+                // modify oldDelta to reflect characters deleted and inserted by an old change
+                oldDelta = oldDelta - oldChange.Span.Length + oldChange.NewLength;
```

Comment:
Did u consider under / overflow scenarios here?

Reply:
I haven't but I'm curious what thoughts you may have on how to mitigate such problems

Reply:
Don't have any great ideas here. More wanted to see if we had considered them sufficiently. This is a bit corner case though. Don't want to block the current fix for extreme cases here.

Reply:
Well, for example, since both `Span.Length` and `NewLength` are required to be positive, does it help prevent overflow if we do `oldDelta = oldDelta + (oldChange.NewLength - oldChange.Span.Length)`? (not going to worry about it for this PR)